### PR TITLE
feat: regex matching for domain names

### PR DESCRIPTION
**Pull request by kate-shine**

### CODE REVIEW

File: synapse_domain_rule_checker/__init__.py [hunk 1]
```diff
@@ -31,20 +32,30 @@ class Membership:
...
 class DomainRuleCheckerConfig:
     can_invite_if_not_in_domain_mapping: bool
-    domain_mapping: Optional[Dict[str, List[str]]] = None
+    use_regex: bool = False
+    domain_mapping: Dict[str, List[str]] = dict()
     can_only_join_rooms_with_invite: bool = False
     can_only_invite_during_room_creation: bool = False
     can_invite_by_third_party_id: bool = True
-    domains_prevented_from_being_invited_to_published_rooms: Optional[List[str]] = None
+    domains_prevented_from_being_invited_to_published_rooms: List[str] = []
```
Comment (jcgruenhage) [inline synapse_domain_rule_checker/__init__.py:41]: Why do these become necessary now, while being optional before?

File: synapse_domain_rule_checker/__init__.py [hunk 2]
```diff
@@ -31,20 +32,30 @@ class Membership:
 @attr.s(auto_attribs=True, frozen=True)
 class DomainRuleCheckerConfig:
     can_invite_if_not_in_domain_mapping: bool
-    domain_mapping: Optional[Dict[str, List[str]]] = None
+    use_regex: bool = False
+    domain_mapping: Dict[str, List[str]] = dict()
```
Comment (jcgruenhage) [inline synapse_domain_rule_checker/__init__.py:37]: same here

File: synapse_domain_rule_checker/__init__.py [hunk 3]
```diff
@@ -172,15 +183,51 @@ async def _user_may_invite(
...
+            and self.match_from_domains(self._domains_prevented_from_being_invited_to_published_rooms, 
+                                        invitee_domain)
         ):
             return False
 
-        if inviter_domain not in self._domain_mapping:
+        if not self.is_domain_mapping_passing(inviter_domain, invitee_domain):
+            return False
+
+        return True
```
Comment (jcgruenhage) [inline synapse_domain_rule_checker/__init__.py]: why not `return self.is_domain_mapping_passing(inviter_domain, invitee_domain)`?

File: synapse_domain_rule_checker/__init__.py [hunk 4]
```diff
@@ -172,15 +183,51 @@ async def _user_may_invite(
...
+        if not self._config.use_regex:
+            return self._domain_mapping.get(inviter_domain)
+
+        matched = False
+        matching = []
+
+        for inviter, invitees in self._domain_mapping.items():
+            if inviter.fullmatch(inviter_domain):
+                matched = True
+                print(invitees)
```
Comment (jcgruenhage) [inline synapse_domain_rule_checker/__init__.py]: Why are you printing here?

**Review by jcgruenhage — COMMENTED** [review]
I'm confused in a few places. Not sure if these *need* changes, but they might, depending on what the answers to my questions are. Also: Will we be upstreaming these changes later on?

File: synapse_domain_rule_checker/__init__.py (hunk 1 above)
Comment (kate-shine) [reply synapse_domain_rule_checker/__init__.py:41]: They are still not necessary, it's just a change of default value if they aren't present. If you take a look at former rows 44-47, those `None` defaults are later changed to empty list and dict anyway.

File: synapse_domain_rule_checker/__init__.py (hunk 4 above)
Comment (kate-shine) [reply synapse_domain_rule_checker/__init__.py]: Smh I searched for loose debug prints, how did I miss one?

File: synapse_domain_rule_checker/__init__.py (hunk 3 above)
Comment (kate-shine) [reply synapse_domain_rule_checker/__init__.py]: Yep, true :)

File: synapse_domain_rule_checker/__init__.py (hunk 1 above)
Comment (jcgruenhage) [reply synapse_domain_rule_checker/__init__.py:41]: Right, those are the default values. Got it.

**Review by jcgruenhage — APPROVED** [review]
LGTM! One tiny nitpick: You're currently signing the commit with a `A3ABDF88541B725C`, which only contains `Kateřina Churanová <k.churanova@famedly.com>`, but the commit has `katerina@churanova.eu` set as the email, which is causing GitHub to be unhappy with the signature. Can you look into that?

**Review by jcgruenhage — APPROVED** [review]

**Comment by jcgruenhage** [comment]
Are we waiting for anything here? I'd like to configure this soon, but we need to merge it first and build a container image with this plugin installed.

**Comment by kate-shine** [comment]
Pls don't delete branch, I will try to PR it to upstream

**Pull request merged**